## Without modules in a non-module project

First, the experiment with an MWE without modules - I think, this should work - or should I activate modules?

MWE - jfxweb.java

```java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 26+
//DEPS org.openjfx:javafx-web:26.0.1
//DEPS org.openjfx:jdk-jsobject:26

import javafx.scene.web.WebView;

public class jfxweb {
  public static void main(String... args) {
	  System.out.println("WebView class: " + WebView.class.getName());
  }
}
```

Output

```terminal
$ jbang jfxweb.java
[jbang] Building jar for jfxweb.java...
Error occurred during initialization of boot layer
java.lang.module.FindException: Module jdk.jsobject not found, required by javafx.web
```

### With module

```java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 26+
//MODULE
//DEPS org.openjfx:javafx-web:26.0.1
//DEPS org.openjfx:jdk-jsobject:26

package mwe;

import javafx.scene.web.WebView;

public class jfxwebmod {
  public static void main(String... args) {
	  System.out.println("WebView class: " + WebView.class.getName());
  }
}
```

Output:

```terminal
$ jbang jfxwebmod.java
[jbang] Building jar for jfxwebmod.java...
C:\TEMP\jfxwebmod.java:9: error: package javafx.scene.web is not visible
import javafx.scene.web.WebView;
                   ^
  (package javafx.scene.web is declared in module javafx.web, but module jfxwebmod does not read it)
1 error
[jbang] [ERROR] Error during compile
[jbang] Run with --verbose or -x for more details. The --verbose or -x must be placed before the jbang command. I.e. jbang --verbose run [...]
```

Much more work seems to be requried to get this running.

---

Claude's longer reasoning - but I think the above says it all

**Disclaimer** Maybe, Claude is totally wrong here, but maybe there is a grain of truth there. Let's see...

For JabRef's `jabkit` (a headless CLI that only drags JavaFX in *transitively* via `jablib` — the user never asked for JavaFX), `//MODULE` hits three separate blockers:

1. **Default package + transitive reads.** The launcher is in the default package, so `//MODULE` first fails with `Module code cannot work with the default package, adding a 'package'
statement is required`. After adding a `package`, compilation fails because the `//SOURCES` import classes from modules pulled in transitively — `afterburner.fx`, `jul.to.slf4j`,
`org.tinylog.api` — none of which the module reads, so each has to be re-declared as an explicit `//DEPS`.

2. **`//MODULE` alone still launches from the class-path.** Without also passing `--module`, `runAsModule` stays false (`JarCmdGenerator`), so JavaFX is on the module-path while
`jdk.jsobject` is on the class-path — same `FindException` as MWE 1.

3. **`jbang run --module` puts *all* deps on the module-path** and then dies on non-modular jars in the tree:

 ```
 java.lang.module.FindException: Unable to derive module descriptor for .../fastparse_2.13-2.3.3.jar
 Caused by: java.lang.IllegalArgumentException: fastparse.2.13: Invalid module name: '2' is not a Java identifier
 ```

 (same for `latex2unicode_2.13`). "Just use modules" can't work once the dependency graph contains any non-modularizable jar, and real-world trees usually do.

### Why this PR instead

It doesn't try to move everything to modules. It stays scoped to the JavaFX case jbang already special-cases: it just keeps the modules the auto-detected JavaFX modules *already*
`requires` (like `jdk.jsobject`) on the same module-path jbang is already building for JavaFX. `--add-modules` is left unchanged.

--- a/src/main/java/dev/jbang/dependencies/ModularClassPath.java
+++ b/src/main/java/dev/jbang/dependencies/ModularClassPath.java
@@ -5,17 +5,13 @@ import static dev.jbang.Settings.CP_SEPARATOR;
 import java.io.File;
 import java.io.IOException;
 import java.net.URI;
-import java.util.ArrayDeque;
 import java.util.ArrayList;
 import java.util.Collections;
-import java.util.Deque;
 import java.util.HashMap;
-import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Optional;
-import java.util.Set;
 import java.util.stream.Collectors;
 
 import org.codehaus.plexus.languages.java.jpms.JavaModuleDescriptor;
@@ -29,6 +25,9 @@ import dev.jbang.util.Util;
 
 public class ModularClassPath {
 	static final String JAVAFX_PREFIX = "javafx";
+	// Required by `javafx.web`; removed from the JDK in 26 and now shipped as
+	// `org.openjfx:jdk-jsobject`.
+	static final String JDK_JSOBJECT_MODULE = "jdk.jsobject";
 
 	private final List<ArtifactInfo> artifacts;
 
@@ -97,52 +96,25 @@ public class ModularClassPath {
 
 				resolvePathsResult.getPathElements().forEach((key, value) -> pathElements.put(key.getPath(), value));
 
-				// Index the resolved modules by name so we can follow `requires` edges.
-				Map<String, String> moduleNameToPath = new HashMap<>();
-				Map<String, JavaModuleDescriptor> moduleNameToDescriptor = new HashMap<>();
-				pathElements.forEach((path, descriptor) -> {
-					if (descriptor != null && descriptor.name() != null) {
-						moduleNameToPath.put(descriptor.name(), path);
-						moduleNameToDescriptor.put(descriptor.name(), descriptor);
+				pathElements.forEach((k, v) -> {
+					if (v != null && v.name() != null
+							&& (v.name().startsWith(JAVAFX_PREFIX) || v.name().equals(JDK_JSOBJECT_MODULE))) {
+						// JavaFX jars belong on the module-path. So does `jdk.jsobject`, which is
+						// required by `javafx.web`: until JDK 25 it was part of the JDK, but JDK 26
+						// removed it and openjfx now ships it as the separate
+						// `org.openjfx:jdk-jsobject`
+						// artifact. Left on the class-path the boot layer cannot find it and fails with
+						// `Module jdk.jsobject not found, required by javafx.web`.
+						// See https://github.com/jbangdev/jbang/issues/559
+						modulePaths.add(k);
+					} else {
+						// classpathElements.add(k);
 					}
 				});
 
-				// Start from the JavaFX modules and transitively collect everything they
-				// require. This way modules like `jdk.jsobject` (required by `javafx.web`)
-				// also end up on the module-path instead of the class-path, where the boot
-				// layer would not find them. See https://github.com/jbangdev/jbang/issues/559
-				Set<String> requiredModules = new HashSet<>();
-				Deque<String> toVisit = new ArrayDeque<>();
-				moduleNameToDescriptor.keySet()
-					.stream()
-					.filter(name -> name.startsWith(JAVAFX_PREFIX))
-					.forEach(toVisit::add);
-				while (!toVisit.isEmpty()) {
-					String name = toVisit.poll();
-					if (!requiredModules.add(name)) {
-						continue;
-					}
-					JavaModuleDescriptor descriptor = moduleNameToDescriptor.get(name);
-					if (descriptor != null) {
-						descriptor.requires()
-							.stream()
-							.map(JavaModuleDescriptor.JavaRequires::name)
-							.filter(moduleNameToDescriptor::containsKey)
-							.forEach(toVisit::add);
-					}
-				}
-
-				// Put the JavaFX modules and their required modules on the module-path.
-				requiredModules.stream()
-					.map(moduleNameToPath::get)
-					.filter(Objects::nonNull)
-					.forEach(modulePaths::add);
-
 				if (!modulePaths.isEmpty()) {
 					commandArguments.add("--module-path");
-					String modulePath = modulePaths.stream()
-						.distinct()
-						.collect(Collectors.joining(File.pathSeparator));
+					String modulePath = String.join(File.pathSeparator, modulePaths);
 					commandArguments.add(modulePath);
 				}
 

